README.md
# Conceal

Simple OpenSSL-based string encryption using a shared secret. The algorithm, initialization vector, salt, crypttext, and HMAC are all encoded into a single string so it is easy to copy around.

## Requirements

* Ruby 1.9.3 or newer

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'conceal'
```

And then execute:

```
$ bundle
```

Or install it yourself as:

```
$ gem install conceal
```

## Usage

This gem provides both a ruby library and some command-line utilities.

### Library

```ruby
encrypted = Conceal.encrypt('some plaintext', key: 'your shared secret', algorithm: 'aes-256-cbc')
decrypted = Conceal.decrypt(encrypted, key: 'your shared secret')
```

### Command-line

```
$ ruby -rsecurerandom -e 'print SecureRandom.urlsafe_base64(32)' | conceal encrypt key.file | pbcopy
$ pbpaste | conceal decrypt key.file | pbcopy
```

## Authors

* Ben Scott (<gamepoet@gmail.com>)


CHANGELOG.md
# Changelog

## 0.2.0

* Backwards incompatible changes
  * Merged `decrypt` and `encrypt` commands into a single binary: `conceal decrypt` and `conceal encrypt`.

## 0.1.0

* Initial release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'conceal'


spec/conceal_spec.rb
require 'spec_helper'

describe Conceal do
  let(:key) { SecureRandom.hex(128) }

  it 'has a version number' do
    expect(Conceal::VERSION).not_to be nil
  end

  it 'encrypt then decrypt returns the same original plaintext' do
    expect(Conceal.decrypt(Conceal.encrypt('hello', key: key), key: key)).to eq('hello')
  end

  describe '#encrypt' do
    it 'does not return the plaintext' do
      expect(Conceal.encrypt('hello', key: key)).not_to eq('hello')
    end

    it 'outputs different values each time (different iv/salt)' do
      first  = Conceal.encrypt('hello', key: key)
      second = Conceal.encrypt('hello', key: key)

      expect(first).not_to eq(second)
    end
  end
end


lib/conceal/cli.rb
require 'thor'

module Conceal
  class CLI < Thor
    map %w(d de dec) => :decrypt
    map %w(e en enc) => :encrypt
    map %w(-v --version) => :version

    option :newline, aliases: '-n', type: :boolean, desc: 'print a trailing newline character'
    desc 'decrypt [OPTIONS] KEY_FILE', 'decrypt stdin using the given key file'
    def decrypt(key_file)
      require 'conceal'

      # load the key
      raise Thor::Error, 'ERROR: key file is not readable or does not exist' unless File.readable?(key_file)
      key = IO.read(key_file)

      # decrypt from stdin
      encrypted_data = $stdin.read
      plaintext = Conceal.decrypt(encrypted_data, key: key)

      $stdout.write(plaintext)
      $stdout.write("\n") if options[:newline]
    end

    option :newline, aliases: '-n', type: :boolean, desc: 'print a trailing newline character'
    desc 'encrypt [OPTIONS] KEY_FILE', 'encrypt stdin using the given key file'
    def encrypt(key_file)
      require 'conceal'

      # load the key
      raise Thor::Error, 'ERROR: key file is not readable or does not exist' unless File.readable?(key_file)
      key = IO.read(key_file)

      # encrypt from stdin
      plaintext = $stdin.read
      encrypted_data = Conceal.encrypt(plaintext, key: key)

      $stdout.write(encrypted_data)
      $stdout.write("\n") if options[:newline]
    end

    desc 'version', 'display the version and exit'
    def version
      require 'conceal/version'
      puts VERSION
    end

  protected
    def self.exit_on_failure?
      true
    end
  end
end


lib/conceal.rb
require 'conceal/version'
require 'base64'
require 'openssl'
require 'securerandom'

module Conceal
  class << self
    FORMAT_VERSION  = 1
    FIELD_SEPARATOR = ':'

    # Encrypts the given plaintext string.
    #
    # @param plaintext  [String]  the plaintext string to encrypt
    # @param opts       [Hash]    additional options
    #
    # @option opts [String] :algorithm  the cipher algorithm to use (defaults to 'aes-25c-cbc')
    # @option opts [String] :key        the secret shared key
    def encrypt(plaintext, opts = {})
      opts = {
        algorithm: 'aes-256-cbc',
      }.merge(opts)
      key       = opts[:key]
      algorithm = opts[:algorithm]
      raise ArgumentError.new(':key option missing') if key.to_s.empty?
      salt = SecureRandom.hex(128)

      # setup the cipher
      cipher = OpenSSL::Cipher::Cipher.new(algorithm)
      cipher.encrypt
      iv = cipher.random_iv
      cipher.iv = iv
      cipher.key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(key, salt, 2000, cipher.key_len)

      # encrypt
      ciphertext = cipher.update(plaintext)
      ciphertext << cipher.final

      # MAC
      digest = OpenSSL::Digest.new('sha256')
      hmac = OpenSSL::HMAC.digest(digest, key, ciphertext)

      [
        FORMAT_VERSION,
        algorithm,
        encode64(iv),
        encode64(salt),
        encode64(hmac),
        encode64(ciphertext),
      ].join(FIELD_SEPARATOR)
    end

    # Decrypts the given encrypted string.
    #
    # @param data [String]    the encrypted string to decrypt
    # @param opts [Hash]      additional options
    #
    # @option opts [String] :key    the secret shared key
    def decrypt(data, opts = {})
      key = opts[:key]
      raise ArgumentError.new(':key option missing') if key.to_s.empty?

      ver, algorithm, iv64, salt64, hmac64, ciphertext64 = data.split(FIELD_SEPARATOR, 6)
      raise ArgumentError.new('ciphertext has unknown version') unless ver == FORMAT_VERSION.to_s

      iv         = Base64.decode64(iv64)
      salt       = Base64.decode64(salt64)
      hmac       = Base64.decode64(hmac64)
      ciphertext = Base64.decode64(ciphertext64)

      # validate the hmac
      digest = OpenSSL::Digest.new('sha256')
      actual_hmac = OpenSSL::HMAC.digest(digest, key, ciphertext)
      raise ArgumentError.new('HMAC mismatch') unless actual_hmac == hmac

      # decrypt
      cipher = OpenSSL::Cipher::Cipher.new(algorithm)
      cipher.decrypt
      cipher.iv = iv
      cipher.key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(key, salt, 2000, cipher.key_len)

      plaintext = cipher.update(ciphertext)
      plaintext << cipher.final
      plaintext
    end

    private
      def encode64(value)
        Base64.encode64(value).gsub(/[\r\n]/, '')
      end
  end
end
